### DISCUSSION TITLE
How to use the Tasks generated by Dynamic Mapping?

### DISCUSSION THREAD

**Discussion by luabida** [apache/airflow#26053] [Q&A]
Hello everyone. I'm implementing Airflow on a project and I've been some struggles when trying to use `expand()` function from python operators. Basically, I have a task that fetches a API and downloads the CSV files via a list comprehension that extracts the `url` from a generator that yields the `table name`s and the `url`s, it is working fine so far.

The problem is that I don't know exactly how many files it will download and I need them to be past forward, so I can operate over each one. I kind of understand the usage of Dynamic Mapping and currently I can see the download task generates 14 other tasks, where each of them carry the argument passed with the function (the csv url) in `op_kwargs`.  I also know a little of the usage of `PythonBranchOperator` and I've been trying use it to create branches from the download task using the url in `op_kwargs` but without any success either.

The workflow will be the same for each task, it should download all the CSVs (working in parallel - nice) >> generate a branch to each CSV (not working) >> [compare csv lines with db:True/False] (the code is working). I just can't figure out the way of using the tasks generated with the `expand()` function. Ty!

**Comment by potiuk** [comment]
You need to generate a task that produces list of individual parameters to be passed to the dynamic tasks and pass it as input to dynamic task's expand. Just follow the examples. Don't use branch operator just do somethign like:

```
 nitial task  -> [list of parameters] -> Dynamic tasks to expand on the list
```

Just understand that you need to have list of metadata objects passed from one task to a dynamic tasks and each element of the list will be passed to a separate instance of the dynamic task. This is a new concept in Airflow but this is notthing new in general - it's mostly how Map-Reduce work and you might also read about Map Reduce concept 

I sugest you simply read and implement the tutorial step by step, play it it with, incrementally adding and playing step-by-step,  try to reason and understand what each step is doing and  https://airflow.apache.org/docs/apache-airflow/2.3.0/concepts/dynamic-task-mapping.html . 

I also might suggest you read original propoasl https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-42+Dynamic+Task+Mapping and try to understand what it is about and what it tries to achieve (and read the comments). 

You might also read some blogs.

Unfortunately this is not a place that you will get hand-holded, people here work contribute for free and spend time rather on writing the docs descrobing the concept and you need to invest a bit of time and learnign (and your own experimentation) to learn it.

**Reply by luabida** [reply]
Hey, thank you for the reply. I could accomplish what I needed with a for loop with the tasks I wanted to be the same for each CSV and seems to be working fine.
